File: 20230702_Sunday/20230702.java
// Leetcode: 1601. Maximum Number of Achievable Transfer Requests.
// https://leetcode.com/problems/maximum-number-of-achievable-transfer-requests/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 72 ms, faster than 18.18% of Java online submissions for Maximum
// Number of Achievable Transfer Requests.
// Memory Usage: 43.6 MB, less than 15.15% of Java online submissions for
// Maximum Number of Achievable Transfer Requests.
// 2023.07.02 Daily Challenge.
class Solution {
    public int maximumRequests(int n, int[][] requests) {
        int answer = 0;

        for (int mask = 0; mask < (1 << requests.length); mask++) {
            int[] indegree = new int[n];
            int pos = requests.length - 1;
            // Number of set bits representing the requests we will consider.
            int bitCount = Integer.bitCount(mask);

            // If the request count we're going to consider is less than the maximum request 
            // We have considered without violating the constraints; then we can return it cannot be the answer.
            if (bitCount <= answer) {
                continue;
            }

            // For all the 1's in the number, update the array indegree for the building it involves.
            for (int curr = mask; curr > 0; curr >>= 1, pos--) {
                if ((curr & 1) == 1) {
                    indegree[requests[pos][0]]--;
                    indegree[requests[pos][1]]++;
                }
            }

            boolean flag = true;
            // Check if it doesn;t violates the constraints
            for (int i = 0; i < n; i++) {
                if (indegree[i] != 0) {
                    flag = false;
                    break;
                }
            }

            if (flag)  {
                answer = bitCount;
            }
        }

        return answer;
    }
}

File: 20230702_Sunday/20230702.rb
# Leetcode: 1601. Maximum Number of Achievable Transfer Requests.
# https://leetcode.com/problems/maximum-number-of-achievable-transfer-requests/
# @param {Integer} n
# @param {Integer[][]} requests
# @return {Integer}
def maximum_requests(n, requests)
  # Based on:
  # https://leetcode.com/problems/maximum-number-of-achievable-transfer-requests/solution/
  ans = 0
  mask = 0
  while (mask < (1 << requests.size)) != 0
    p mask
    indegree = Array.new(n,0)
    pos = requests.size - 1
    bit_count = mask.digits(2).count(1)
    if bit_count <= ans
      mask += 1
      next
    end

    curr = mask
    while curr > 0
      indegree[requests[pos][0]] -= 1
      indegree[requests[pos][1]] += 1
      # # # # #
      curr >>= 1
      pos -= 1
    end

    flag = true
    (0...n).each do |i|
      next if indegree[i] == 0
      flag = false
      break
    end

    answer = bit_count if flag

    # # # # #
    mask += 1
  end
  return ans
end
